mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -117,6 +132,276 @@ struct AdjointOpPattern : public OpConversionPattern<AdjointOp> {
+        LLVMTypeConverter llvmTypeConverter(ctx);
+
+        Type vectorType =
+            llvmTypeConverter.convertType(MemRefType::get({UNKNOWN}, Float64Type::get(ctx)));
+
+        for (Type type : op.getResultTypes()) {
+            if (!type.isa<MemRefType>())
+                return op.emitOpError("must be bufferized before lowering");
+        }
+
+        // The callee of the backprop Op
+        func::FuncOp callee =
+            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+        assert(callee);
+
+        // Creat constants
+        Value c0 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI32IntegerAttr(0));
+        Value c1 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI32IntegerAttr(1));
+
+        // Create mlir memref to llvm
+        StringRef allocFnName = "_mlir_memref_to_llvm_alloc";
+        Type allocFnSignature = LLVM::LLVMFunctionType::get(LLVM::LLVMPointerType::get(ctx),
+                                                            {IntegerType::get(ctx, 64)});
+
+        LLVM::LLVMFuncOp allocFnDecl =
+            ensureFunctionDeclaration(rewriter, op, allocFnName, allocFnSignature);
+        // Create the Enzyme function
+        StringRef backpropFnName = "__enzyme_autodiff";
+        Type backpropFnSignature =
+            LLVM::LLVMFunctionType::get(LLVM::LLVMVoidType::get(ctx), {}, /*isVarArg=*/true);
+
+        LLVM::LLVMFuncOp backpropFnDecl =
+            ensureFunctionDeclaration(rewriter, op, backpropFnName, backpropFnSignature);
+
+        // Create the memset function
+        StringRef memsetFnName = "memset";
+        Type memsetFnSignature =
+            LLVM::LLVMFunctionType::get(LLVM::LLVMPointerType::get(ctx),
+                                        {LLVM::LLVMPointerType::get(ctx), IntegerType::get(ctx, 32),
+                                         IntegerType::get(ctx, 64)});
+
+        LLVM::LLVMFuncOp memsetFnDecl =
+            ensureFunctionDeclaration(rewriter, op, memsetFnName, memsetFnSignature);
+
+        // Generate the wrapper function for argmapfn
+        func::FuncOp wrapper = genEnzymeWrapperFunction(rewriter, loc, callee);
+
+        // Set up the first argument and transforms the wrapper value in ptr
+        FunctionType wrapperType = wrapper.getFunctionType();
+        Value wrapperValue =
+            rewriter.create<func::ConstantOp>(loc, wrapperType, wrapper.getName()).getResult();
+        Type wrapperPtrType = llvmTypeConverter.convertType(wrapperType);
+        Value wrapperPtr =
+            rewriter.create<UnrealizedConversionCastOp>(loc, wrapperPtrType, wrapperValue)
+                .getResult(0);
+
+        // Add the pointer to the wrapped callee
+        SmallVector<Value> callArgs = {wrapperPtr};
+
+        // Add the arguments and their shadow on data in
+        for (auto [arg, llvmarg, llvmdatain] :
+             llvm::zip(op.getArgs(), adaptor.getArgs(), adaptor.getDataIn())) {
```

Comment:
I'm a bit confused by the names here. What is an `arg` vs a `dataIn`?

Reply:
It was decided a while back to remove calls to `malloc` in the runtime. For operations that returned a tensor, we have this convention where they receive an argument called `DataIn` (I supposed, one could rename it to `DataOut`). The lowering is roughly:

```
%res = foo.op() -> tensor<...>
```

After bufferization:

```
%memref = memref.alloc
foo.op %memref : memref<...> 
```

and memref here is the name of the data in.

If you see ways in which this can be improved, happy to listen and take suggestions!

Reply:
Destination passing style for quantum operations that return tensors